[libstdc++] initial integration

pauldreik opened the pull request:
this adds a fuzzer for format in the gnu standard C++ library

this particular fuzzer has found several bugs which have been reported upstream and fixed. ([110860](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=110860), [110862](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=110862) and [110868](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=110868))

the code is partially based on the fuzzers I made for the {fmt} library already on oss-fuzz.

I have contacted Jonathan Wakely @jwakely from upstream who has agreed to be a mail cc and is OK with this work taking place.

since fuzzing a standard library while using another one is a bit weird, I am not sure I got everything right but this is a place to start and improve from. in particular, memory sanitizer does not work so I disabled it.

jwakely commented:
> I have contacted Jonathan Wakely @jwakely from upstream who has agreed to be a mail cc and is OK with this work taking place.

Just confirming what Paul said. I'm very OK with this fuzzing and hope to expand it to cover other parts of libstdc++ too.

I'm the lead maintainer for libstdc++ so please use me as the upstream contact.

jwakely commented on projects/libstdcpp/Dockerfile:
```diff
@@ -0,0 +1,24 @@
+RUN git clone --depth 1 git://gcc.gnu.org/git/gcc.git libstdcpp
+WORKDIR libstdcpp
+COPY build.sh $SRC/
+ADD https://raw.githubusercontent.com/pauldreik/stdfuzz/main/format/one-arg.cpp $SRC/
```
I think the regex fuzzers mostly just hit implementation limits due to excessive memory usage or stack depths and so segfault. We already have about a dozen bugs about those limits, so constantly running fizzers to re-find them isn't helpful, and will just generate a lot of noise.

pauldreik commented:
@MarcoFalke 
The idea is to start small and get this running with a single fuzzer to begin with, so we get the machinery in place. It needs to get accepted into oss-fuzz first! 
Eventually, we can add more fuzzers and move the code to upstream but my experience is that it good to be flexible in the beginning.
I looked at libc++, which has the fuzzers integrated upstream and the script here just invokes the checked in script. That is what I want as well, but let's start small and see what the oss-fuzz maintainers say!

The one-arg.cpp fuzzer is just for a subset of the format api. There is more to do on format. For instance, I expect formatting of std::chrono types to be error prone.

Alan32Liu commented:
Hi @pauldreik thanks for the PR!
There seems to be a CI failure with fuzzer `Centipede`, could you please fix it or exclude `Centipede`?

We can exclude a fuzzer by explicitly including other fuzzers in `project.yaml`:
```yaml
fuzzing_engines:
  - libfuzzer
  - afl
  - honggfuzz
  # - centipede
  ```

pauldreik commented:
> Hi @pauldreik thanks for the PR! There seems to be a CI failure with fuzzer `Centipede`, could you please fix it or exclude `Centipede`?
> 
> We can exclude a fuzzer by explicitly including other fuzzers in `project.yaml`:
> 
> ```yaml
> fuzzing_engines:
>   - libfuzzer
>   - afl
>   - honggfuzz
>   # - centipede
> ```

Thanks! Added that change, squashed and force pushed.

Alan32Liu commented:
> Thanks! Added that change, squashed and force pushed.

Thanks a lot!
I will enable auto-merge so that it can be merged as soon as we pass all CIs : )
